### PR TITLE
Elasticsearch Plugin

### PR DESCRIPTION
**Pull request by mfleader**
## Changes introduced with this PR

Load data into an elasticsearch instance.

---
By contributing to this repository, I agree to the [contribution guidelines](https://github.com/arcalot/.github/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,22 @@
+[tool.poetry]
+name = "arcaflow-plugin-elasticsearch"
+version = "0.1.0"
```
Comment (sandrobonazzola) [inline pyproject.toml:3]: you can set this to `"0.0.0"` and have github action updating this on tagged build.
See: https://github.com/arcalot/arcaflow-plugin-sdk-python/blob/main/pyproject.toml#L8
and https://github.com/arcalot/arcaflow-plugin-sdk-python/blob/main/.github/workflows/build.yml#L60

File: es_plugin.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    except BaseException:
+        return "error", ErrorOutput(
+            "failed"
+        )
+
+
+if __name__ == "__main__":
+    sys.exit(plugin.run(plugin.build_schema(
+        run,
+    )))
```
Comment (sandrobonazzola) [inline es_plugin.py]: missing new line at EOF

File: es_plugin.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+def run(
+
+) -> typing.Tuple[str, typing.Union[SuccessOutput, ErrorOutput]]:
+    """
+
+    :return: the string identifying which output it is,
+             as well the output structure
+    """
+    try:
+        pass
```
Comment (sandrobonazzola) [inline es_plugin.py]: looks like the implementation is missing :-)

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+ENTRYPOINT ["python3", "/app/es_plugin.py"]
+CMD []
+
+LABEL org.opencontainers.image.source="https://github.com/arcalot/arcaflow-plugin-elasticsearch"
+LABEL org.opencontainers.image.licenses="Apache-2.0+GPL-2.0-only"
+LABEL org.opencontainers.image.vendor="Arcalot project"
+LABEL org.opencontainers.image.authors="Arcalot contributors"
+LABEL org.opencontainers.image.title="Python Elasticsearch Plugin"
+LABEL io.github.arcalot.arcaflow.plugin.version="1"
```
Comment (sandrobonazzola) [inline Dockerfile:26]: missing new line at EOF

File: .github/workflows/carpenter.yaml [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+  carpenter-build:
+    name: build
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
+      - name: carpenter build
+        uses: arcalot/arcaflow-plugin-image-builder@main
+        with:
+          args: build --build --push
```
Comment (sandrobonazzola) [inline .github/workflows/carpenter.yaml:25]: missing new line at EOF

**Review by sandrobonazzola — CHANGES_REQUESTED** [review]
Inline comments

File: es_schema.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    index: Annotated[str, validation.min(1)] = field(
+        metadata={
+            'name': 'index',
+            'description': """Name of the index that """
+            """will receive this data."""
+        }
+    )
+    data: Annotated[typing.Optional[schema.ANY_TYPE]] = field(
+        metadata={
+            'name': 'data',
```
Comment (portante) [inline es_schema.py:61]: Lots of questions come up with this plug-in.

Typically one does not just index one document into Elasticsearch.  You have a series of N documents that you want indexed.  So while this is fine for an initial step, we need to discuss the nature of this as a plugin.

FWIW, we wrote [pyesbulk](https://pypi.org/project/pyesbulk/) to enable an efficient way of indexing a stream of data into Elasticsearch, with error handling, and able to handle multiple documents.

It is primarily used by the [pbench server code](https://github.com/distributed-system-analysis/pbench/blob/665132a898e068a6c488610ccc447fd33db68035/lib/pbench/server/indexer.py#L180).

Comment (ghost) [reply es_schema.py:61]: We don't have data streams in Arcaflow yet, hence we need to call one by one. Once streams are supported we can consider adding that support.

Comment (portante) [reply es_schema.py:61]: One document at a time isn't going to be worth it in a workflow, you'll either have a huge scaling problem trying to describe all the individual documents in the workflow, or you'll have too large a document to be useful in Elasticsearch.

**Comment by portante** [comment]
Having a reason for closing a PR with unresolved conversations would be helpful.  Is this because it is superseded by #2?

**Pull request closed without merging**